src/docs/services/relay/projectconfig-versioning.mdx

```diff
@@ -0,0 +1,138 @@
+    SpanOperations { ... },
+    #[serde(other)]
+    Unsupported,
+}
+```
+
+Before adding a new variant in Sentry, verify that Relay implements this correctly. If this is not the case, proceed by replacing the field just like when changing a fields type. Never add variants to an enum that does not have an unknown variant.
+
+It is acceptable that Relay does not forward unknown enum variants by dropping the original value, as it is done with the above approach.
+
+## Adding Top-Level Fields
+
+In case you're developing a new feature, you likely want to add completely new data to the project config and not touch existing structs. Here's a small checklist that applies in addition to everything else:
+
+- Verify that the field is in `camelCase`.
+- Ensure the field name has not been used in a past version of Relay. `git log -G field_name` is a simple way to search git history for strings that occur in diffs.
+- Make sure your field is either optional or has a default
+- Do not serialize when the field is `None` or default via `serde(skip_serializing_if)`.
+- Use `ErrorBoundary` around the field type for as long as your new feature is pre-GA and can change its definition. This makes sure that if the field definition is malformed, Relay contains the error and continues to parse the rest of the project config.
+- Do not use `usize`.
+
+It can be tempting to use `usize` if for some reason the max value of your field is platform-dependent (i.e. whether the Relay that deserializes currently runs on a 32-bit OS). This *can be* the case if your field is an index into some array/vec. Even in those cases it is barely worth it. Keep in mind that not only the current Relay needs to use that number as array index, but also downstream Relays.
+
+## Adding Subfields
+
+In principle, it's permitted to add new fields at any time. Relay ignores additional fields, so struct field definitions can be implemented and deployed at a later point. For example, a struct like this will deserialize just fine from a JSON message with additional fields:
+
+```rust
+struct Foo {
+    foo: u64,
+}
+
+// compatible with: {"foo": 123}, {"foo": 123, "bar": 456}, {"foo": 123, ...}
+// incompatible with: {"foo": "123"}
+```
+
+There are two caveats:
+
+- As demonstrated above, changing types is *not* safe. See below for instructions on changing types.
+- Ensure the field name has never been used before. It could have been implemented with different types, or trigger unwanted behavior in prior versions of Relay.
+- Note that adding enum variants requires a different approach.
+
+## Renaming Fields
+
+Do not rename a field for features that are generally available for external Relays. Renaming a field makes the field invisible to older Relays. This makes it infeasible to rename fields for aesthetic reasons (it *may* be fine if your feature is pre-EA), but for the same reason [it is a useful mechanism for versioning](#breaking-changes-and-actual-versioning)
+
+## Changing Field Types
```

Comment:
I think we need to revise this section:

1. Let's lead with external Relays, because this is the critical case. You can then lower the constraints for internal Relays. 
2. Type changes are only permitted when they have exactly identical serialization behavior. For example, changing `String` to `Cow<str>` would be allowed. There are no other exceptions, everything else is forbidden.

Types cannot be widened, not even with untagged enums. This is because Relay would then forward the new value to a downstream that doesn't support it yet. 

Types cannot be made stricter, because then Relay might reject project configs from an older onpremise version.

Reply:
>There are no other exceptions, everything else is forbidden.

I think enlarging integer types is probably safe most of the time but I just added a general disclaimer that this needs to be discussed with the team

Applied the rest of the feedback